shared/src/balancer/graph_api.rs

```diff
@@ -0,0 +1,413 @@
+    }
+}
+
+mod pools_query {
+    use crate::balancer::{pool_storage::RegisteredWeightedPool, swap::fixed_point::Bfp};
+    use anyhow::{anyhow, ensure, Result};
+    use ethcontract::{H160, H256};
+    use serde::Deserialize;
+
+    pub const QUERY: &str = r#"
+        query Pools($block: Int, $pageSize: Int, $lastId: ID) {
+            pools(
+                block: { number: $block }
+                first: $pageSize
+                where: {
+                    id_gt: $lastId
+                    poolType: Weighted
+                }
+            ) {
+                id
+                address
+                tokens {
+                    address
+                    decimals
+                    weight
+                }
+                historicalValues(
+                    first: 1
+                    orderBy: block
+                ) {
+                    block
+                }
+            }
+        }
+    "#;
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    pub struct Data {
+        pub pools: Vec<Pool>,
+    }
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Pool {
+        pub id: H256,
+        pub address: H160,
+        pub tokens: Vec<Token>,
+        pub historical_values: Vec<HistoricalValue>,
+    }
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    pub struct Token {
+        pub address: H160,
+        pub decimals: u8,
+        #[serde(with = "serde_with::rust::display_fromstr")]
+        pub weight: Bfp,
+    }
+
+    #[derive(Debug, Deserialize, PartialEq)]
+    pub struct HistoricalValue {
+        #[serde(with = "serde_with::rust::display_fromstr")]
+        pub block: u64,
+    }
+
+    impl Pool {
+        pub fn into_registered(self, block_fetched: u64) -> Result<RegisteredWeightedPool> {
+            // The Balancer subgraph does not contain information for the block
+            // in which a pool was created. Instead, we assume it is the block
+            // that the data was fetched or the oldest historical liquidity
+            // values, depending on whether the latter is available.
+            let block_created_guesstimate = self
+                .historical_values
+                .get(0)
+                .map(|value| value.block)
+                .unwrap_or(block_fetched);
+            ensure!(
+                block_created_guesstimate <= block_fetched,
+                "historical data more recent than fetched block",
+            );
+
+            let token_count = self.tokens.len();
+            self.tokens.iter().try_fold(
+                RegisteredWeightedPool {
+                    pool_id: self.id,
+                    pool_address: self.address,
+                    tokens: Vec::with_capacity(token_count),
+                    normalized_weights: Vec::with_capacity(token_count),
+                    scaling_exponents: Vec::with_capacity(token_count),
+                    block_created: block_created_guesstimate,
```

Comment:
block created only matters for reorgs, doesn't it? Can't we just use block_fetched in that case if we ensure this is never older than our reorg threshold?

Reply:
So my initial intuition was that we could use this historic liquidity data to get closer to the actual block created. That way, if there were for whatever reason a deep re-org it wouldn't really affect the pools (especially the "main" ones that have existed presumably for a long time).

However, with your suggestion of using the event handlers reorg threshold, this is no longer an issue as the event updater would error because of this instead of losing all the pools. I like this much better.